rfcs/51-animation-composition.md

```diff
@@ -0,0 +1,434 @@
+
+### Graph Storage
+Two main stores are needed internally for the graph: node storage and clip
+storage. Nodes are stored in a flat `Vec<Node>`, and a newtyped `NodeId` is used
+to keep a reference to a node within the graph. Due to this structure, nodes
+cannot be removed from the graph. As a worksaround, a node can be effectively
+removed by disconnecting all of it's incoming and outgoing edges. The root NodeId
+is always 0.
+
+Clip storage decomposes clips based on the properties each clip is animating.
+Instead of using `Handle<AnimationClip>`, each clip is instead internally
+assigned a auto-imcrementing `ClipId` much like `NodeId`. However, instead of
+storing a `Vec<AnimationClip`, a map of property paths to `Track` is stored.
+
+`Track` stores a contiguous sparse array of pointers to all curves associated
+with a given property. Here's a rough outline of how such a struct might look:
+
+```rust
+struct VariableCurve<T: Animatable> {
+  keys: Vec<T>,
+  times: Vec<f32>,
+}
+
+enum Track {
+  Translation(Vec<Option<Arc<VariableCurve<Vec3A>>>>),
+  Scale(Vec<Option<Arc<VariableCurve<Vec3A>>>>),
+  Rotation(Vec<Option<Arc<VariableCurve<Quat>>>>),
+  // Optional. Potential representation for reflection based animation.
+  Reflect(Vec<Option<Arc<VariableCurve<Box<dyn Animatable>>>>>),
+}
+```
+
+The individual curves stored within are indexed by `ClipId`, and each pointer to
+a curve is non-None if and only if the original clip had a curve for the
+corresponding property.
+
+### Graph Evaluation
+```rust
+struct Influence {
+    weight: f32,
+    time: f32,
+}
+
+struct GraphInfluences {
+    influence: Vec<Influence> // Indexed by BoneId
+}
+```
+
+To remove the need to evaluate the graph every time a property is sampled, an
+influences map is constructed based on the current state of the graph. This is
+effectively a `Vec<f32>` indexed on `ClipId` mapping clips and their respective
+*cumulative* weights. To avoid constant reallocation, this influences map is
+stored as a companion component and is cleared before every evaluation.
+
+During evaluation, the graph is traversed along every connected edge and weights
+are multiplicatively propagated downward. For example, if we have a simple graph
+`A -> B -> C`, where A is the root and C is the final clip and the edges `A -> B`
+and `B -> C` have the weights 0.5 and 0.25 respectively, the final weight for
+clip C is 0.125. If multiple paths to a end clip are present, the final weight
+for the clip is the sum of all path's weights.
+
+After traversing the graph, the weights of all active inputs are clamped and
+normalized.
+
+By default, a dedicated system will run before sampling every app tick that
+evaluates every changed graph, relying on change detection to ensure that any
+mutation to the graph results in a change in the influences. This behavior can be
+disabled via a config option, requiring users to manually evaluate the graph
+instead.
+
+### Binding
+
+```rust
+struct BoneId(usize);
```

Comment:
The animation system animates more than "bones", already with morph targets https://github.com/bevyengine/bevy/pull/8158 animations are not limited to skinned mesh bones. I'd like a more general wording.